src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -537,6 +555,217 @@ def create_certificate_signing_request(self, cert_info):
+                        f'DNS Challenge not found for domain {authorization_resource.body.identifier.value}'
+                    )
+                    continue
+
+                try:
+                    status = self.middleware.call_sync(
+                        'dns.authenticator.update_txt_record', {
+                            'authenticator': domain_names_dns_mapping[domain],
+                            'challenge': challenge.json_dumps(),
+                            'domain': domain,
+                            'key': key.json_dumps()
+                        }
+                    )
+                except ValidationErrors as v:
+                    verrors.extend(v)
+                else:
+                    try:
+                        acme_client.answer_challenge(challenge, challenge.response(key))
+                    except errors.UnexpectedUpdate as e:
+                        verrors.add(
+                            'acme_authorization.domain',
+                            f'Error answering challenge for {domain} : {e}'
+                        )
+
+            finally:
+                job.set_progress(
+                    progress,
+                    f'DNS challenge {"completed" if status else "failed"} for {domain}'
+                )
+
+        if verrors:
+            raise verrors
+
+    @periodic(86400, run_on_start=True)
+    @job(lock='acme_cert_renewal')
+    def renew_certs(self, job):
+        certs = self.middleware.call_sync(
+            'certificate.query',
+            [['acme', '!=', None]]
+        )
+
+        progress = 0
+        for cert in certs:
+            progress += (100 / len(certs))
+
+            if (
+                    datetime.datetime.strptime(cert['until'], '%a %b %d %H:%M:%S %Y') -
+                    datetime.datetime.utcnow()
+            ).days < cert['renew_days']:
+                # renew cert
+                self.logger.debug(f'Renewing certificate {cert["name"]}')
+                final_order = self.acme_issue_certificate(
+                    job, progress / 4, {
+                        'tos': True,
+                        'acme_directory_uri': cert['acme']['directory'],
+                        'dns_mapping': cert['domains_authenticators']
+                    },
+                    cert
+                )
+
+                self.middleware.call_sync(
+                    'datastore.update',
+                    self._config.datastore,
+                    cert['id'],
+                    {
+                        'certificate': final_order.fullchain_pem,
+                        'acme_uri': final_order.uri,
+                        'chain': True if len(RE_CERTIFICATE.findall(final_order.fullchain_pem)) > 1 else False,
+                    },
+                    {'prefix': self._config.datastore_prefix}
+                )
+
+            job.set_progress(progress)
+
+    @accepts()
+    async def popular_acme_server_choices(self):
```

Comment:
popular doesnt read good for the API user. Lets just call it `acme_server_choices`.

Let also make it return a dict
  key: value

e.g.

'https://acme-staging-v02.api.letsencrypt.org/directory': 'Lets Encrypt Staging'